### ISSUE TITLE
Duplicate classes when used codegenModels.set("responseBased")

### ISSUE THREAD

**Issue by alexxxdev** [apollographql/apollo-kotlin#4562]
### Summary

When using codegenModels.set("responseBased"), i get an error after generating classes:
```Conflicting declarations: public final data class Image : ImageData, public final data class Image : ImageData```

### Version

3.7.1

### Logs

```
query GetFavorites {
    favorites{
        list {
            id
            image {
                ...imageData
            }
            images {
                ...imageData
            }
        }
    }
}

fragment imageData on Image {
    id
}

////

public data class List(
  public val id: Int?,
  public val image: Image?,
  public val images: kotlin.collections.List<Image?>?,
) {
  public fun imagesFilterNotNull(): kotlin.collections.List<Image>? = images?.filterNotNull()

  public data class Image(
    public val __typename: String,
    public override val id: Int?,
  ) : ImageData {
    public companion object {
      public fun Image.imageData() = this as? ImageData
    }
  }

  public data class Image(
    public val __typename: String,
    public override val id: Int?,
  ) : ImageData {
    public companion object {
      public fun Image.imageData() = this as? ImageData
    }
  }
}
```

**Comment by BoD** [comment]
Hi! 👋 

The reason this happens is that the codegen generates a class for each composite field, named after the field's name - but for lists, the name is made singular first. In your case `images` will be generated as a `List<Image>`. But there is also an `image` field in the query which will also generate a singular `Image` type, hence the conflict.

The conflict can be avoided by setting an alias to either field, the generated type will be named after it. For instance:

```graphql
            singleImage: image {
                ...imageData
            }
            images {
                ...imageData
            }
```

**Comment by alexxxdev** [comment]
thanks, this works, but it seems less convenient to use
and yet, it seems to me that this seems like a bug

> Hi! 👋
> 
> The reason this happens is that the codegen generates a class for each composite field, named after the field's name - but for lists, the name is made singular first. In your case `images` will be generated as a `List<Image>`. But there is also an `image` field in the query which will also generate a singular `Image` type, hence the conflict.
> 
> The conflict can be avoided by setting an alias to either field, the generated type will be named after it. For instance:
> 
> ```graphql
>             singleImage: image {
>                 ...imageData
>             }
>             images {
>                 ...imageData
>             }
> ```

**Comment by martinbonnin** [comment]
> I understand that the solution to the current problem is not trivial and I'm trying to find a solution in other ways

Yep, the `responseBased` codegen is quite complex. For the record, this problem happens when there both a single and list field with the same name (`image` and `images` in your case). The list model will be singularized (because it would be weird to have `List<Images>`). When that happens, your best bet is to use aliases as [described in this comment above](https://github.com/apollographql/apollo-kotlin/issues/4562#issuecomment-1342406792).

I'll tweak the name of this issue to add auto-escaping of model names in that case. I think we should go with `Image` and `Image1` to make it consistent with `flattenModels.set(true)`.
